bionic/task_state.py

```diff
@@ -0,0 +1,306 @@
+            self._results_by_name = results_by_name
+
+        return results_by_name
+
+    def _compute(self, task_key_logger):
+        """
+        Computes the values of a task state by running its task. Requires that all
+        the task's dependencies are already complete.
+        """
+
+        task = self.task
+
+        dep_results = [
+            dep_state.get_results(task_key_logger)[dep_key.dnode.to_entity_name()]
+            for dep_state, dep_key in zip(self.dep_states, task.dep_keys)
+        ]
+
+        provider = self.provider
+
+        if not task.is_simple_lookup:
+            for task_key in task.keys:
+                task_key_logger.log_computing(task_key)
+
+        dep_values = [dep_result.value for dep_result in dep_results]
+
+        values = task.compute(dep_values)
+        assert len(values) == len(provider.attrs.names)
+
+        for query in self.queries:
+            if task.is_simple_lookup:
+                task_key_logger.log_accessed_from_definition(query.task_key)
+            else:
+                task_key_logger.log_computed(query.task_key)
+
+        results_by_name = {}
+        result_value_hashes_by_name = {}
+        for ix, (query, value) in enumerate(zip(self.queries, values)):
+            query.protocol.validate(value)
+
+            result = Result(query=query, value=value,)
+
+            if provider.attrs.should_persist():
+                accessor = self.cache_accessors[ix]
+                accessor.save_result(result)
+
+                value_hash = accessor.load_result_value_hash()
+                result_value_hashes_by_name[query.dnode.to_entity_name()] = value_hash
+
+            results_by_name[query.dnode.to_entity_name()] = result
+
+        # Memoize results at this point only if results should not persist.
+        # Otherwise, load it lazily later so that if the serialized/deserialized
+        # value is not exactly the same as the original, we still
+        # always return the same value.
+        if provider.attrs.should_memoize() and not provider.attrs.should_persist():
+            self._results_by_name = results_by_name
+
+        # But we cache the hashed values eagerly since they are cheap to load.
+        if provider.attrs.should_persist():
+            self.result_value_hashes_by_name = result_value_hashes_by_name
+
+    def sync_after_subprocess_completion(self):
+        """
+        Syncs the task state by populating and reloading data in the current process
+        after completing the task state in a subprocess.
+
+        Values populated in the task state are not communicated back from the subprocess.
+        """
+
+        # First, reload cache accessors
+        for accessor in self.cache_accessors:
+            accessor.flush_stored_entries()
+
+        # Then, populate the value hash
+        if self.result_value_hashes_by_name is None:
+            self.result_value_hashes_by_name = {}
+            for ix, query in enumerate(self.queries):
```

Comment:
You could avoid using an index with `for query, accessor in zip(self.queries, self.cache_accessors)`. But it would be even simpler to do this, I think:

    for accessor in self.cache_accessors:
        value_hash = accessor.load_result_value_hash()
        self.result_value_hashes_by_name[
            accessor.query.dnode.to_entity_name()
        ] = value_hash

Reply:
Oh accessor contains query too. That makes it easy.